gremlin-python/src/main/python/gremlin_python/driver/auth.py

```diff
@@ -0,0 +1,81 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+import abc
+
+
+class Auth(metaclass=abc.ABCMeta):
```

Comment:
Following the Java implementation here seems a bit odd to me. The Auth interface only has one function apply() which makes it like a functional interface. This is needed in Java because functions aren't first class. The same isn't true for Python. In gremlin-python, you should just accept a list of functions that take in one parameter for the request. This function would be the interceptor and should be allowed to change the request as needed. The auth implementation should use that same underlying interceptor mechanism to update the rquest.

Reply:
Updated